### ISSUE TITLE
Cloning entities

### ISSUE THREAD

**Issue by kentcb** [BrightstarDB/BrightstarDB#174]
Is there a built-in mechanism for cloning entities? I can't seem to find one, nor can I find docs on this.

Use case: I have an entity that I'm about to delete, but afterwards I need to notify observers that the entity has been deleted. Deleting the entity is resetting dates to `DateTime.MinValue`, so I can't just use the same entity instance that is passed into `DeleteObject`.

Seems like cloning logic could be generated too?

**Comment by kal** [comment]
I'm not sure about adding cloning support - it would mean pushing a requirement on to developers of partial class extensions to ensure that they also have cloning support in there. I wonder if there is a different way of handling your use case though as I can see how it would be a useful thing to have.

**Comment by kentcb** [comment]
Good point about entity extensibility. The way I dealt with this is to change my entity->contract extension method helpers to produce immutable values rather than simply wrapping the entity and exposing it via the contract interface. This entails deep cloning, but it's done at a natural point in the code base. Not only does this solve the problem, but I think it's just a better design anyway because now all data from my services is immutable.

Tying this back to the use case I mentioned, I simply call `ToContract()` on the entity *before* I delete it. Thus, I have a deep clone of it that conforms to my contract which I can then use to notify listeners after a successful deletion.

Feel free to close this if you like.

**Comment by kal** [comment]
I'm going to leave it open as a reminder that it would be good to have some story on post-delete property accesses - it seems reasonable to expect to be able to access the properties of an entity after it has been deleted from the store for as long as you are holding on to the entity handle.

**Comment by kal** [comment]
Going to bump this out of the 1.11 release as it might be better to think about it in the context of adding Deleted / Deleting delegates to the BrightstarEntityContext.
